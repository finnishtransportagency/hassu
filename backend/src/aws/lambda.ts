import { Lambda } from "@aws-sdk/client-lambda";
import { log } from "../logger";

const lambda: Lambda = new Lambda({});

export async function invokeLambda(functionName: string, synchronousCall: boolean, payload?: string): Promise<string | undefined> {
  try {
    if (synchronousCall) {
      const output = await lambda.invoke({
        FunctionName: functionName,
        Payload: new TextEncoder().encode(payload || "{}"),
      });
      const outputPayload = output.Payload;
      return new TextDecoder("utf-8").decode(outputPayload);
    } else {
      await lambda.invokeAsync({
        FunctionName: functionName,
        InvokeArgs: payload || "{}",
      });
    }
  } catch (e) {
    log.error("invokeLambda", e);
    throw e;
  }
}
