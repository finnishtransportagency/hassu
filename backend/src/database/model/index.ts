export * from "./common";
export * from "./projekti";
export * from "./suunnitteluVaihe";
export * from "./nahtavillaoloVaihe";
export * from "./hyvaksymisPaatosVaihe";
