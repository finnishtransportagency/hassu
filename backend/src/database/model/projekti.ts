import { Status } from "../../api/apiModel";

export type DBProjekti = {
  oid: string;
  nimi?: string | null;
  vaylamuoto?: string[];
  vaihe?: string | null;
  tila?: string | null;
  asianumero?: string | null;
  tyyppi?: string | null;
  status?: Status | null;
  organisaatio?: string | null;
  toteutusAjankohta?: string | null;
  vastuuhenkilo?: string | null;
  // false, jos projekti ladattiin Velhosta, mutta ei ole viel√§ tallennettu tietokantaan
  tallennettu?: boolean;
};
