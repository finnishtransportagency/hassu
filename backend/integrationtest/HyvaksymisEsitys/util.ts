import { config } from "../../src/config";
import { getS3Client } from "../../src/aws/client";
import { DeleteObjectsCommand, ListObjectsV2Command, ListObjectsV2CommandOutput, PutObjectCommand } from "@aws-sdk/client-s3";
import { createPresignedPost } from "@aws-sdk/s3-presigned-post";
import { adaptFileName } from "../../src/tiedostot/paths";
import { getAxios } from "../../src/aws/monitoring";
import FormData from "form-data";

export async function insertYllapitoFileToS3(pathInS3: string) {
  if (pathInS3.includes(config.yllapitoBucketName)) {
    throw new Error("path ei saa sisältää ylläpitoBucketNamea");
  }
  const buffer = Buffer.from("Test");
  await getS3Client().send(
    new PutObjectCommand({
      Body: buffer,
      Bucket: config.yllapitoBucketName,
      Key: pathInS3,
    })
  );
}

export async function insertUploadFileToS3(uuid: string, filename: string) {
  const fileNameWithPath = `${uuid}/${adaptFileName(filename)}`;
  const presignedPost = await createPresignedPost(getS3Client(), {
    Bucket: config.uploadBucketName,
    Key: fileNameWithPath,
    Expires: 600,
  });
  const fields = presignedPost.fields;
  const form = new FormData();
  Object.keys(fields).forEach((key) => {
    form.append(key, fields[key]);
  });
  form.append("file", "Test");
  const axios = getAxios();
  const lengthSync = form.getLengthSync();

  await axios.post(presignedPost.url, form, {
    headers: {
      "Content-Length": lengthSync,
      ...form.getHeaders(),
    },
  });
}

export async function deleteYllapitoFiles(sourcePrefix?: string) {
  if (sourcePrefix && sourcePrefix.includes(config.yllapitoBucketName)) {
    throw new Error("sourcePrefix ei saa sisältää yllapitoBucketNamea");
  }
  const s3 = getS3Client();
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.yllapitoBucketName,
        Prefix: sourcePrefix,
        ContinuationToken,
      })
    );
    await deleteFiles(Contents.map(({ Key }) => Key as string));

    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
}

export async function emptyUploadFiles() {
  const s3 = getS3Client();
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.uploadBucketName,
        ContinuationToken,
      })
    );
    await deleteFiles(
      Contents.map(({ Key }) => Key as string),
      config.uploadBucketName
    );

    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
}

async function deleteFiles(paths: string[], bucket: string = config.yllapitoBucketName) {
  if (paths.some((path) => path.includes(config.yllapitoBucketName))) {
    throw new Error("path ei saa sisältää ylläpitoBucketNamea");
  }
  await Promise.all(
    // DeleteObjectsCommand takes at most 1000 items
    getChunksOfThousand(paths).map((paths) =>
      getS3Client().send(
        new DeleteObjectsCommand({
          Bucket: bucket,
          Delete: { Objects: paths.map((path) => ({ Key: path })) },
        })
      )
    )
  );
}

function getChunksOfThousand<T>(array: T[]): T[][] {
  const chunks: T[][] = [];
  for (let i = 0; i < array.length; i += 1000) {
    chunks.push(array.slice(i, i + 1000));
  }
  return chunks;
}

export async function getYllapitoFilesUnderPath(pathInS3: string): Promise<string[]> {
  const s3 = getS3Client();
  const files: string[] = [];
  let ContinuationToken = undefined;
  do {
    const { Contents = [], NextContinuationToken }: ListObjectsV2CommandOutput = await s3.send(
      new ListObjectsV2Command({
        Bucket: config.yllapitoBucketName,
        Prefix: pathInS3,
        ContinuationToken,
      })
    );
    files.push(...Contents.map(({ Key }) => Key as string));
    ContinuationToken = NextContinuationToken;
  } while (ContinuationToken);
  return files;
}
