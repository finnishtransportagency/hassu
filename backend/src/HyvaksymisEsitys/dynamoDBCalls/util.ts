import { UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { log } from "../../logger";
import { getDynamoDBDocumentClient } from "../../aws/client";
import { ConditionalCheckFailedException } from "@aws-sdk/client-dynamodb";
import { SimultaneousUpdateError } from "hassu-common/error";

export async function sendUpdateCommandToDynamoDB(params: UpdateCommand): Promise<void> {
  if (log.isLevelEnabled("debug")) {
    log.debug("Updating projekti to Hassu with params", { params });
  }
  try {
    const dynamoDBDocumentClient = getDynamoDBDocumentClient();
    await dynamoDBDocumentClient.send(params);
  } catch (e) {
    if (e instanceof ConditionalCheckFailedException) {
      throw new SimultaneousUpdateError("Projektia on p√§ivitetty tietokannassa. Lataa projekti uudelleen.");
    }
    log.error(e instanceof Error ? e.message : String(e), { params });
    throw e;
  }
}
