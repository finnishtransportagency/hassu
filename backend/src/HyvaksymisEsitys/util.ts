import { IllegalArgumentError } from "hassu-common/error";
import { DBProjekti } from "../database/model";
import { projektiDatabase } from "../database/projektiDatabase";
import { requirePermissionLuku } from "../user";
import { NykyinenKayttaja } from "hassu-common/graphql/apiModel";

export async function varmistaLukuoikeusJaHaeProjekti(
  oid: string
): Promise<{ projektiInDB: DBProjekti; nykyinenKayttaja: NykyinenKayttaja }> {
  const nykyinenKayttaja = requirePermissionLuku();
  const projektiInDB = await projektiDatabase.loadProjektiByOid(oid);
  if (!projektiInDB) {
    throw new IllegalArgumentError(`Projektia oid:lla ${oid} ei l√∂ydy`);
  }
  return { projektiInDB, nykyinenKayttaja };
}
