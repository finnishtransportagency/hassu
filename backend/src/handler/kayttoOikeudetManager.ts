import { DBVaylaUser } from "../database/model/projekti";
import { ProjektiKayttaja, ProjektiKayttajaInput, ProjektiRooli } from "../../../common/graphql/apiModel";
import { personSearch, SearchMode } from "../personSearch/personSearchClient";
import * as log from "loglevel";
import differenceWith from "lodash/differenceWith";

export class KayttoOikeudetManager {
  private users: DBVaylaUser[];

  constructor(users: DBVaylaUser[]) {
    this.users = users;
  }

  async applyChanges(changes: ProjektiKayttajaInput[] | undefined | null) {
    if (!changes) {
      return;
    }
    // Go through all users in database projekti and the input.
    // Update existing ones, remove those that are missing in input, and add those that exist only in input.
    const resultUsers: DBVaylaUser[] = this.users.reduce((resultingUsers: DBVaylaUser[], currentUser) => {
      const inputUser = changes.find((user) => user.kayttajatunnus === currentUser.kayttajatunnus);
      if (inputUser) {
        // Update only puhelinnumero if projektipaallikko
        if (inputUser.rooli === ProjektiRooli.PROJEKTIPAALLIKKO) {
          resultingUsers.push({
            ...currentUser,
            puhelinnumero: inputUser.puhelinnumero,
          });
        } else {
          // Update rest of fields
          resultingUsers.push({
            ...currentUser,
            ...inputUser,
          });
        }
      } else {
        // Remove user because it doesn't exist in input, except projektipaallikko which cannot be removed
        if (currentUser.rooli === ProjektiRooli.PROJEKTIPAALLIKKO) {
          resultingUsers.push(currentUser);
        }
      }
      return resultingUsers;
    }, []);

    // Add new users
    const newUsers = differenceWith(changes, resultUsers, (u1, u2) => u1.kayttajatunnus === u2.kayttajatunnus);
    await Promise.all(
      newUsers.map(async (newUser) => {
        const userToAdd = {
          puhelinnumero: newUser.puhelinnumero,
          kayttajatunnus: newUser.kayttajatunnus,
          rooli: newUser.rooli,
        } as any;
        try {
          const userWithAllInfo = await personSearch.fillInUserInfoFromUserManagement({
            user: userToAdd,
            searchMode: SearchMode.UID,
          });
          if (userWithAllInfo) {
            resultUsers.push(userWithAllInfo);
          }
        } catch (e) {
          log.error(e);
        }
      })
    );
    this.users = resultUsers;
  }

  getKayttoOikeudet() {
    return this.users;
  }

  getAPIKayttoOikeudet(): ProjektiKayttaja[] {
    return this.users.map((user) => ({
      __typename: "ProjektiKayttaja",
      ...user,
    }));
  }

  async addProjektiPaallikkoFromEmail(vastuuhenkiloEmail: string) {
    return await this.resolveProjektiPaallikkoFromVelhoVastuuhenkilo(vastuuhenkiloEmail);
  }

  async resolveProjektiPaallikkoFromVelhoVastuuhenkilo(vastuuhenkiloEmail: string) {
    // Replace or create new projektipaallikko
    this.removeProjektiPaallikko();
    const projektiPaallikko = await personSearch.fillInUserInfoFromUserManagement({
      user: { rooli: ProjektiRooli.PROJEKTIPAALLIKKO, email: vastuuhenkiloEmail, esitetaanKuulutuksessa: true } as any,
      searchMode: SearchMode.EMAIL,
    });
    if (projektiPaallikko) {
      this.users.push(projektiPaallikko);
      return projektiPaallikko;
    }
  }

  async addUserByKayttajatunnus(kayttajatunnus: string, rooli: ProjektiRooli) {
    const user = await personSearch.fillInUserInfoFromUserManagement({
      user: { kayttajatunnus, rooli } as any,
      searchMode: SearchMode.UID,
    });
    if (user) {
      this.users.push(user);
      return user;
    }
  }

  private removeProjektiPaallikko() {
    this.users = this.users.filter((user) => user.rooli === ProjektiRooli.PROJEKTIPAALLIKKO);
  }
}
