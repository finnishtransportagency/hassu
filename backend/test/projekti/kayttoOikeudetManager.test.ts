import { describe, it } from "mocha";
import { KayttoOikeudetManager } from "../../src/projekti/kayttoOikeudetManager";
import { PersonSearchFixture } from "../personSearch/lambda/personSearchFixture";
import { Kayttajas } from "../../src/personSearch/kayttajas";
import { ELY, KayttajaTyyppi, ProjektiKayttajaInput } from "../../../common/graphql/apiModel";
import { DBVaylaUser } from "../../src/database/model";
import { organisaatioIsEly } from "../../src/util/organisaatioIsEly";

const { expect } = require("chai");

describe("KayttoOikeudetManager", () => {
  let kayttajas: Kayttajas;

  const personSearchFixture: PersonSearchFixture = new PersonSearchFixture();
  const vaylaKayttajaA1 = personSearchFixture.createKayttaja("A1");
  const elyKayttajaA2 = personSearchFixture.createKayttaja("A2", "ELY");
  const vaylaKayttajaA3 = personSearchFixture.createKayttaja("A3");
  const elyKayttajaA4 = personSearchFixture.createKayttaja("A4", "ELY");
  const konsulttiKayttajaLX5 = personSearchFixture.createKayttaja("LX5", "ramboll");

  let users: DBVaylaUser[];

  beforeEach(() => {
    kayttajas = Kayttajas.fromKayttajaList([vaylaKayttajaA1, elyKayttajaA2, vaylaKayttajaA3, elyKayttajaA4, konsulttiKayttajaLX5]);
    users = [
      {
        tyyppi: KayttajaTyyppi.PROJEKTIPAALLIKKO,
        kayttajatunnus: vaylaKayttajaA1.uid as string,
        email: vaylaKayttajaA1.email as string,
        etunimi: vaylaKayttajaA1.etunimi as string,
        sukunimi: vaylaKayttajaA1.sukunimi as string,
        organisaatio: vaylaKayttajaA1.organisaatio as string,
        puhelinnumero: vaylaKayttajaA1.puhelinnumero as string,
        muokattavissa: false,
      },
      {
        tyyppi: KayttajaTyyppi.VARAHENKILO,
        kayttajatunnus: elyKayttajaA2.uid as string,
        email: elyKayttajaA2.email as string,
        etunimi: elyKayttajaA2.etunimi as string,
        sukunimi: elyKayttajaA2.sukunimi as string,
        organisaatio: elyKayttajaA2.organisaatio as string,
        puhelinnumero: elyKayttajaA2.puhelinnumero as string,
        muokattavissa: false,
      },
      {
        tyyppi: null,
        kayttajatunnus: vaylaKayttajaA3.uid as string,
        email: vaylaKayttajaA3.email as string,
        etunimi: vaylaKayttajaA3.etunimi as string,
        sukunimi: vaylaKayttajaA3.sukunimi as string,
        organisaatio: vaylaKayttajaA3.organisaatio as string,
        puhelinnumero: vaylaKayttajaA3.puhelinnumero as string,
        muokattavissa: true,
      },
    ];
  });

  function expectProjektiPaallikko(manager: KayttoOikeudetManager, uid: string) {
    const projektiPaallikko = manager
      .getKayttoOikeudet()
      .filter((user) => user.tyyppi == KayttajaTyyppi.PROJEKTIPAALLIKKO)
      .filter((user) => user.kayttajatunnus == uid)
      .pop();
    expect(projektiPaallikko).to.exist;
    expect(projektiPaallikko!.muokattavissa).to.be.false;
  }

  function expectVarahenkilo(manager: KayttoOikeudetManager, uid: string, muokattavissa: boolean) {
    const varahenkilo = manager
      .getKayttoOikeudet()
      .filter((user) => user.tyyppi == KayttajaTyyppi.VARAHENKILO)
      .filter((user) => user.kayttajatunnus == uid)
      .pop();
    expect(varahenkilo, "varahenkilöä ei löydy: " + [uid, muokattavissa, JSON.stringify(manager.getKayttoOikeudet())].join(", ")).to.exist;
    expect(varahenkilo!.muokattavissa).to.eq(muokattavissa);
  }

  function expectKayttaja(manager: KayttoOikeudetManager, uid: string, data: Partial<DBVaylaUser>) {
    const varahenkilo = manager
      .getKayttoOikeudet()
      .filter((user) => user.kayttajatunnus == uid)
      .pop();
    expect(varahenkilo, "henkilöä ei löydy: " + [uid, JSON.stringify(manager.getKayttoOikeudet())].join(", ")).to.exist;
    expect(varahenkilo).to.include(data);
  }

  it("should manager projekti users successfully", async () => {
    const manager = new KayttoOikeudetManager([], kayttajas);

    // Lisää projektipäällikkö Velhosta
    manager.addProjektiPaallikkoFromEmail(vaylaKayttajaA1.email);
    expectProjektiPaallikko(manager, "A1");

    // Lisää varahenkilö Velhosta
    manager.addVarahenkiloFromEmail(elyKayttajaA2.email);
    expectProjektiPaallikko(manager, "A1");
    expectVarahenkilo(manager, "A2", false);

    // Lisää nykyinen käyttäjä varahenkilöksi
    manager.addUser({ kayttajatunnus: vaylaKayttajaA3.uid!, muokattavissa: true, tyyppi: KayttajaTyyppi.VARAHENKILO });
    expectProjektiPaallikko(manager, "A1");
    expectVarahenkilo(manager, "A2", false);
    expectVarahenkilo(manager, "A3", true);

    // Lisää varahenkilo velhosta, jolla ei ole A- tai L-tunnusta
    manager.addVarahenkiloFromEmail(konsulttiKayttajaLX5.email);
    expectKayttaja(manager, "LX5", { tyyppi: null, muokattavissa: false });

    // Muokkaa sisältöä ja lisää käyttäjä käyttöliittymältä tulevan pyynnön perusteella
    expect(manager.getKayttoOikeudet()).toMatchSnapshot();
    manager.applyChanges([
      {
        kayttajatunnus: vaylaKayttajaA1.uid!,
        puhelinnumero: vaylaKayttajaA1.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA2.uid!,
        puhelinnumero: elyKayttajaA2.puhelinnumero!,
      },
      {
        kayttajatunnus: vaylaKayttajaA3.uid!,
        puhelinnumero: vaylaKayttajaA3.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA4.uid!,
        puhelinnumero: elyKayttajaA4.puhelinnumero!,
        yleinenYhteystieto: true,
      },
    ]);

    expectKayttaja(manager, "A1", { puhelinnumero: vaylaKayttajaA1.puhelinnumero! });
    expectKayttaja(manager, "A2", { puhelinnumero: elyKayttajaA2.puhelinnumero! });
    expectKayttaja(manager, "A3", { puhelinnumero: vaylaKayttajaA3.puhelinnumero! });
    expectKayttaja(manager, "A4", { puhelinnumero: elyKayttajaA4.puhelinnumero! });
    expect(manager.getKayttoOikeudet()).to.have.length(5);
    expect(manager.getKayttoOikeudet()).toMatchSnapshot();

    // Vaihda käyttäjäA4 varahenkilöksi
    manager.applyChanges([
      {
        kayttajatunnus: vaylaKayttajaA1.uid!,
        puhelinnumero: vaylaKayttajaA1.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA2.uid!,
        puhelinnumero: elyKayttajaA2.puhelinnumero!,
      },
      {
        kayttajatunnus: vaylaKayttajaA3.uid!,
        puhelinnumero: vaylaKayttajaA3.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA4.uid!,
        puhelinnumero: elyKayttajaA4.puhelinnumero!,
        tyyppi: KayttajaTyyppi.VARAHENKILO,
      },
    ]);

    expectKayttaja(manager, "A4", { tyyppi: KayttajaTyyppi.VARAHENKILO });

    // Tarkista, ettei tyyppiä voi vaihtaa projektipäälliköksi
    const kayttoOikeudetbefore = manager.getKayttoOikeudet();
    // Yritä vaihtaa käyttäjäA4 projektipäälliköksi
    manager.applyChanges([
      {
        kayttajatunnus: vaylaKayttajaA1.uid!,
        puhelinnumero: vaylaKayttajaA1.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA2.uid!,
        puhelinnumero: elyKayttajaA2.puhelinnumero!,
      },
      {
        kayttajatunnus: vaylaKayttajaA3.uid!,
        puhelinnumero: vaylaKayttajaA3.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA4.uid!,
        puhelinnumero: elyKayttajaA4.puhelinnumero!,
        tyyppi: KayttajaTyyppi.PROJEKTIPAALLIKKO,
      },
    ]);
    // Tarkista, ettei mitään muuttunut
    expect(manager.getKayttoOikeudet()).to.eql(kayttoOikeudetbefore);
    manager.applyChanges(null);
    expect(manager.getKayttoOikeudet()).to.eql(kayttoOikeudetbefore);

    // Yritä poistaa projektipäällikkö ja varahenkilö
    manager.applyChanges([
      {
        kayttajatunnus: vaylaKayttajaA3.uid!,
        puhelinnumero: vaylaKayttajaA3.puhelinnumero!,
      },
      {
        kayttajatunnus: elyKayttajaA4.uid!,
        puhelinnumero: elyKayttajaA4.puhelinnumero!,
        tyyppi: KayttajaTyyppi.PROJEKTIPAALLIKKO,
      },
    ]);
    // Tarkista, ettei mitään muuttunut
    expect(manager.getKayttoOikeudet()).to.eql(kayttoOikeudetbefore);
  });

  it("should promote existing user to projektipäällikkö successfully", async () => {
    const manager = new KayttoOikeudetManager([], kayttajas);

    // Lisää nykyinen käyttäjä
    manager.addUser({ kayttajatunnus: vaylaKayttajaA3.uid!, muokattavissa: true });
    // Lisää sama käyttäjä Velhosta projektipäälliköksi
    manager.addProjektiPaallikkoFromEmail(vaylaKayttajaA3.email);

    const kayttoOikeudet = manager.getKayttoOikeudet();
    expect(kayttoOikeudet).to.have.length(1);
    const resultUser = kayttoOikeudet[0];
    expect(resultUser).to.eql({
      tyyppi: KayttajaTyyppi.PROJEKTIPAALLIKKO,
      kayttajatunnus: "A3",
      muokattavissa: false,
      organisaatio: "Väylävirasto",
      email: "a3@vayla.fi",
      etunimi: "EtunimiA3",
      sukunimi: "SukunimiA3",
    });
  });

  it("should promote existing user to varahenkilö successfully", async () => {
    const manager = new KayttoOikeudetManager([], kayttajas);

    // Lisää nykyinen käyttäjä
    manager.addUser({ kayttajatunnus: vaylaKayttajaA3.uid!, muokattavissa: true });
    // Lisää sama käyttäjä Velhosta varahenkilöksi
    manager.addVarahenkiloFromEmail(vaylaKayttajaA3.email);

    const kayttoOikeudet = manager.getKayttoOikeudet();
    expect(kayttoOikeudet).to.have.length(1, JSON.stringify(kayttoOikeudet, null, 2));
    const resultUser = kayttoOikeudet[0];
    expect(resultUser).to.eql({
      tyyppi: KayttajaTyyppi.VARAHENKILO,
      kayttajatunnus: "A3",
      muokattavissa: false,
      organisaatio: "Väylävirasto",
      email: "a3@vayla.fi",
      etunimi: "EtunimiA3",
      sukunimi: "SukunimiA3",
    });
  });

  it("should not allow kunnanEdustaja to be removed when applying changes", async () => {
    const manager = new KayttoOikeudetManager(users, kayttajas, vaylaKayttajaA3.uid || undefined);
    manager.applyChanges([
      {
        kayttajatunnus: vaylaKayttajaA1.uid as string,
        puhelinnumero: vaylaKayttajaA1.puhelinnumero as string,
      },
      {
        kayttajatunnus: elyKayttajaA2.uid as string,
        puhelinnumero: elyKayttajaA2.puhelinnumero as string,
      },
    ]);
    const kayttoOikeudet = manager.getKayttoOikeudet();
    expect(kayttoOikeudet.length).eql(3);
  });

  it("should not allow kunnanEdustaja to be removed when doing addProjektiPaallikkoFromEmail", async () => {
    const manager = new KayttoOikeudetManager(users, kayttajas, vaylaKayttajaA1.uid || undefined);
    manager.addProjektiPaallikkoFromEmail(elyKayttajaA4.email);
    const kayttoOikeudet = manager.getKayttoOikeudet();
    expect(kayttoOikeudet.length).eql(4);
  });

  it("should not allow kunnanEdustaja to be removed when doing addVarahenkiloFromEmail", async () => {
    const manager = new KayttoOikeudetManager(users, kayttajas, elyKayttajaA2.uid || undefined);
    manager.addVarahenkiloFromEmail(elyKayttajaA4.email);
    const kayttoOikeudet = manager.getKayttoOikeudet();
    expect(kayttoOikeudet.length).eql(4);
  });

  it("should modify elyOrganisaatio only for users in organisation 'ELY'", async () => {
    const manager = new KayttoOikeudetManager(users, kayttajas);
    const muutoksetElyorganisaatioilla: ProjektiKayttajaInput[] = [
      { kayttajatunnus: vaylaKayttajaA1.uid!, puhelinnumero: vaylaKayttajaA1.puhelinnumero!, elyOrganisaatio: ELY.HAME_ELY },
      { kayttajatunnus: elyKayttajaA2.uid!, puhelinnumero: elyKayttajaA2.puhelinnumero!, elyOrganisaatio: ELY.HAME_ELY },
      { kayttajatunnus: vaylaKayttajaA3.uid!, puhelinnumero: vaylaKayttajaA3.puhelinnumero!, elyOrganisaatio: ELY.HAME_ELY },
      { kayttajatunnus: elyKayttajaA4.uid!, puhelinnumero: elyKayttajaA4.puhelinnumero!, elyOrganisaatio: ELY.HAME_ELY },
    ];
    manager.applyChanges(muutoksetElyorganisaatioilla);
    const kayttoOikeudet = manager.getKayttoOikeudet();
    kayttoOikeudet.forEach((kayttaja) => {
      if (organisaatioIsEly(kayttaja.organisaatio)) {
        expect(kayttaja.elyOrganisaatio).eql(ELY.HAME_ELY);
      } else {
        expect(kayttaja.elyOrganisaatio).eql(undefined);
      }
    });
  });
});
