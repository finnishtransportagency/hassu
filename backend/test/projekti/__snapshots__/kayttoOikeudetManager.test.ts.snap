// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KayttoOikeudetManager should manager projekti users successfully 1`] = `
Array [
  Object {
    "email": "a1@vayla.fi",
    "etunimi": "EtunimiA1",
    "kayttajatunnus": "A1",
    "muokattavissa": false,
    "organisaatio": "Väylävirasto",
    "sukunimi": "SukunimiA1",
    "tyyppi": "PROJEKTIPAALLIKKO",
    "yleinenYhteystieto": true,
  },
  Object {
    "email": "a2@organisaatio.fi",
    "etunimi": "EtunimiA2",
    "kayttajatunnus": "A2",
    "muokattavissa": false,
    "organisaatio": "ELY",
    "sukunimi": "SukunimiA2",
    "tyyppi": "VARAHENKILO",
  },
  Object {
    "email": "a3@vayla.fi",
    "etunimi": "EtunimiA3",
    "kayttajatunnus": "A3",
    "muokattavissa": true,
    "organisaatio": "Väylävirasto",
    "sukunimi": "SukunimiA3",
    "tyyppi": "VARAHENKILO",
  },
  Object {
    "email": "lx5@organisaatio.fi",
    "etunimi": "EtunimiLX5",
    "kayttajatunnus": "LX5",
    "muokattavissa": false,
    "organisaatio": "ramboll",
    "sukunimi": "SukunimiLX5",
    "tyyppi": null,
  },
]
`;

exports[`KayttoOikeudetManager should manager projekti users successfully 2`] = `
Array [
  Object {
    "elyOrganisaatio": undefined,
    "email": "a1@vayla.fi",
    "etunimi": "EtunimiA1",
    "kayttajatunnus": "A1",
    "muokattavissa": false,
    "organisaatio": "Väylävirasto",
    "puhelinnumero": "123456789",
    "sukunimi": "SukunimiA1",
    "tyyppi": "PROJEKTIPAALLIKKO",
    "yleinenYhteystieto": true,
  },
  Object {
    "elyOrganisaatio": undefined,
    "email": "a2@organisaatio.fi",
    "etunimi": "EtunimiA2",
    "kayttajatunnus": "A2",
    "muokattavissa": false,
    "organisaatio": "ELY",
    "puhelinnumero": "123456789",
    "sukunimi": "SukunimiA2",
    "tyyppi": "VARAHENKILO",
    "yleinenYhteystieto": false,
  },
  Object {
    "email": "a3@vayla.fi",
    "etunimi": "EtunimiA3",
    "kayttajatunnus": "A3",
    "muokattavissa": true,
    "organisaatio": "Väylävirasto",
    "puhelinnumero": "123456789",
    "sukunimi": "SukunimiA3",
    "tyyppi": "VARAHENKILO",
  },
  Object {
    "email": "lx5@organisaatio.fi",
    "etunimi": "EtunimiLX5",
    "kayttajatunnus": "LX5",
    "muokattavissa": false,
    "organisaatio": "ramboll",
    "sukunimi": "SukunimiLX5",
    "tyyppi": null,
  },
  Object {
    "elyOrganisaatio": undefined,
    "email": "a4@organisaatio.fi",
    "etunimi": "EtunimiA4",
    "kayttajatunnus": "A4",
    "muokattavissa": true,
    "organisaatio": "ELY",
    "puhelinnumero": "123456789",
    "sukunimi": "SukunimiA4",
    "tyyppi": undefined,
    "yleinenYhteystieto": true,
  },
]
`;
